### ISSUE TITLE
 rollout/advantages_mean=nan, rollout/advantages_max=nan, rollout/advantages_min=nan,

### ISSUE THREAD

**Issue by airrain** [RLinf/RLinf#273]
log :
[1:41:04<128:04:18, 467.13s/it, time/time/sync_weights=0.607, time/time/generate_rollouts=302, time/time/cal_adv_and_returns=0.0159, time/time/actor_training=165, time/time/step=467, env/success_once=0.375, env/return=0.3125, env/episode_len=160.0, env/reward=0.001953125, rollout/rewards=0.00195, rollout/advantages_mean=nan, rollout/advantages_max=nan, rollout/advantages_min=nan, train/actor/approx_kl=0.0211, train/actor/clip_fraction=0, train/actor/clipped_ratio=0.923, train/actor/dual_cliped_ratio=0, train/actor/grad_norm=nan, train/actor/lr=5e-6, train/actor/policy_loss=nan,

config:
defaults:

    env/train: libero_goal
    env/eval: libero_goal
    override hydra/job_logging: stdout

hydra:
run:
dir: .
output_subdir: null
searchpath:
- file://${oc.env:EMBODIED_PATH}/config/

cluster:
num_nodes: 1
component_placement:
actor: 1-2
env: 0-0
rollout: 3-3

runner:
task_type: embodied
logger:
log_path: "../results"
project_name: rlinf
experiment_name: "test_openpi"
logger_backends: ["tensorboard"] # wandb, swanlab

max_epochs: 1000
max_steps: -1

val_check_interval: -1
save_interval: 40
seq_length: 512
max_prompt_length: 25

algorithm:
auto_reset: False
ignore_terminations: False
use_fixed_reset_state_ids: True
normalize_advantages: True
kl_penalty: kl # how to estimate kl divergence: kl or kl_penalty
group_size: 1
max_episode_steps: 160
use_ordered_reset_state_ids: False
reward_coef: 1.0
n_chunk_steps: 32
n_eval_chunk_steps: 64
num_group_envs: 2
rollout_epoch: 8
reward_type: chunk_level
logprob_type: chunk_level
entropy_type: token_level

update_epoch: 4
adv_type: grpo
loss_type: actor
loss_agg_func: "token-mean"
kl_beta: 0.0
entropy_bonus: 0
clip_ratio_high: 0.2
clip_ratio_low: 0.2
clip_ratio_c: 3.0
value_clip: 0.2
huber_delta: 10.0

gamma: 0.99
gae_lambda: 0.95

filter_rewards: False
rewards_lower_bound: 0.1
rewards_upper_bound: 0.9
params for generation

sampling_params:
use_greedy: False
temperature_train: 1.0
temperature_eval: 0.6
top_k: 50
top_p: 1.0
repetition_penalty: 1.0
add_BOS: False
length argument for autoregressive sampling
max length means max amount of tokens to generate

length_params:
max_new_token: null
max_length: 512
min_length: 1

env:
group_name: "EnvGroup"
channel:
name: "env_buffer_list"
queue_name: "obs_buffer"
queue_size: 0
enable_offload: False

rollout:
group_name: "RolloutGroup"
channel:
name: ${env.channel.name}
queue_name: "action_buffer"
queue_size: 0
mode: "colocate"
generation_backend: "huggingface"
model_dir: "/home/midea/vla_projects/RLinf-main/RLinf-Pi0-SFT-Spatial-Object-Goal"
precision: ${actor.model.precision}
recompute_logprobs: False
unnorm_key: libero_10
enable_offload: False
pipeline_stage_num: 1

actor:
group_name: "ActorGroup"
channel:
name: ${env.channel.name}
queue_name: "replay_buffer"
queue_size: 0
training_backend: "fsdp"
checkpoint_load_path: "/home/midea/vla_projects/RLinf-main/RLinf-Pi0-SFT-Spatial-Object-Goal"
checkpoint_save_path: "../results"
micro_batch_size: 1
global_batch_size: 4 # maximum : global batch size = micro batch size * actor num_processes.
seed: 42
enable_offload: False

model:
model_name: "openpi"
precision: null
num_action_chunks: 5 # interface for the env
action_dim: 7
is_lora: False
lora_rank: 2
gradient_checkpointing: False
use_wrist_image: True
use_proprio: True
num_steps: 4
# openpi specific parameters
openpi:
noise_level: 0.5
action_chunk: ${actor.model.num_action_chunks}
num_steps: ${actor.model.num_steps}
train_expert_only: True
action_env_dim: ${actor.model.action_dim}
noise_method: "flow_sde"
add_value_head: False
detach_critic_input: False

optim:
lr: 5.0e-6
value_lr: 1.0e-4
adam_beta1: 0.9
adam_beta2: 0.95
adam_eps: 1.0e-05
clip_grad: 1.0

fsdp_config: # below are fsdp configs, for more details, see https://pytorch.org/docs/stable/fsdp.html and https://huggingface.co/docs/accelerate/package_reference/fsdp

strategy: "fsdp"            # FSDP strategy: ["fsdp", "fsdp2"]

# Sharding strategy: "full_shard" (shard parameters, gradients, optimizer states),
# "shard_grad_op" (shard gradients and optimizer states only), 
# "hybrid_shard" (combines data parallelism and model parallelism - sharding within FSDP groups, replicate across DDP groups),
# "no_shard" (no sharding)
sharding_strategy: "no_shard"

cpu_offload: True           # whether to offload parameters and gradients to CPU when not in use，if True, actor's enable_offload should be True too.
offload_pin_memory: False    # whether FSDP2's CPU offload policy should pin memory when cpu_offload is True
reshard_after_forward: True  # if True, FSDP2 will reshard parameters after forward pass to save memory

enable_gradient_accumulation: True  # if True, gradient accumulation will be enabled for FSDP/FSDP2 training, which will enhance training performance but increase memory usage.
forward_prefetch: False             # if True, FSDP will explicitly prefetches the next upcoming all-gather while executing in the forward pass. only use with static graphs. Overlaps communication with computation to improve performance.
limit_all_gathers: False            # if True, FSDP will synchronizes CPU threads to limit the number of concurrent all-gathers. Only affects strategies that schedule all-gathers
backward_prefetch: null             # options are null, 'pre', 'post'. if 'pre', FSDP will prefetch the next upcoming all-gather while computing gradients. if 'post', FSDP will prefetch the next all-gather until current gradient is computed.
use_orig_params: False              # if True, FSDP will use module's original parameters, it means it will expose nn.Module.named_parameters rather than FlatParameter
use_liger_kernel: False             # if True, liger_kernel will be used for FSDP, note that currently supported models in RLinf are [qwen2.5, qwen2.5-vl], for more details, see liger_kernel's doc

fsdp_size: -1               # Number of GPUs per FSDP group for hybrid sharding. -1 means use all available GPUs in a single FSDP group

mixed_precision:            # mixed precision settings for fsdp/fsdp2
  param_dtype: ${actor.model.precision}
  reduce_dtype: ${actor.model.precision}
  buffer_dtype: ${actor.model.precision}

amp:
  enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
  precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
  use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.

reward:
use_reward_model: False

critic:
use_critic_model: False

**Comment by zoeyuchao** [comment]
have u changed any config in yaml？

**Comment by airrain** [comment]
log  : 
 [1:41:04<128:04:18, 467.13s/it, time/time/sync_weights=0.607, time/time/generate_rollouts=302, time/time/cal_adv_and_returns=0.0159, time/time/actor_training=165, time/time/step=467, env/success_once=0.375, env/return=0.3125, env/episode_len=160.0, env/reward=0.001953125, rollout/rewards=0.00195, rollout/advantages_mean=nan, rollout/advantages_max=nan, rollout/advantages_min=nan, train/actor/approx_kl=0.0211, train/actor/clip_fraction=0, train/actor/clipped_ratio=0.923, train/actor/dual_cliped_ratio=0, train/actor/grad_norm=nan, train/actor/lr=5e-6, train/actor/policy_loss=nan,

config: 
defaults:
  - env/train: libero_goal
  - env/eval: libero_goal
  - override hydra/job_logging: stdout

hydra:
  run:
    dir: .
  output_subdir: null
  searchpath:
    - file://${oc.env:EMBODIED_PATH}/config/

cluster:
  num_nodes: 1
  component_placement:
    actor: 1-2
    env: 0-0
    rollout: 3-3

runner:
  task_type: embodied
  logger:
    log_path: "../results"
    project_name: rlinf
    experiment_name: "test_openpi"
    logger_backends: ["tensorboard"] # wandb, swanlab

  max_epochs: 1000
  max_steps: -1

  val_check_interval: -1
  save_interval: 40
  seq_length: 512
  max_prompt_length: 25

algorithm:
  auto_reset: False
  ignore_terminations: False
  use_fixed_reset_state_ids: True
  normalize_advantages: True
  kl_penalty: kl  # how to estimate kl divergence: kl or kl_penalty
  group_size: 1
  max_episode_steps: 160
  use_ordered_reset_state_ids: False
  reward_coef: 1.0
  n_chunk_steps: 32
  n_eval_chunk_steps: 64
  num_group_envs: 2
  rollout_epoch: 8
  reward_type: chunk_level
  logprob_type: chunk_level
  entropy_type: token_level

  update_epoch: 4
  adv_type: grpo
  loss_type: actor
  loss_agg_func: "token-mean" 
  kl_beta: 0.0
  entropy_bonus: 0
  clip_ratio_high: 0.2
  clip_ratio_low: 0.2
  clip_ratio_c: 3.0
  value_clip: 0.2
  huber_delta: 10.0

  gamma: 0.99
  gae_lambda: 0.95

  filter_rewards: False
  rewards_lower_bound: 0.1
  rewards_upper_bound: 0.9
  # params for generation
  sampling_params:
    use_greedy: False
    temperature_train: 1.0
    temperature_eval: 0.6
    top_k: 50
    top_p: 1.0
    repetition_penalty: 1.0
    add_BOS: False

  # length argument for autoregressive sampling
  # max length means max amount of tokens to generate
  length_params:
    max_new_token: null
    max_length: 512
    min_length: 1

env:
  group_name: "EnvGroup"
  channel:
    name: "env_buffer_list"
    queue_name: "obs_buffer"
    queue_size: 0
  enable_offload: False

rollout:
  group_name: "RolloutGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "action_buffer"
    queue_size: 0
  mode: "colocate"
  generation_backend: "huggingface"
  model_dir: "/home/midea/vla_projects/RLinf-main/RLinf-Pi0-SFT-Spatial-Object-Goal"
  precision: ${actor.model.precision}
  recompute_logprobs: False
  unnorm_key: libero_10
  enable_offload: False
  pipeline_stage_num: 1

actor:
  group_name: "ActorGroup"
  channel:
    name: ${env.channel.name}
    queue_name: "replay_buffer"
    queue_size: 0
  training_backend: "fsdp"
  checkpoint_load_path: "/home/midea/vla_projects/RLinf-main/RLinf-Pi0-SFT-Spatial-Object-Goal"
  checkpoint_save_path: "../results"
  micro_batch_size: 1
  global_batch_size: 4 # maximum : global batch size = micro batch size * actor num_processes.
  seed: 42
  enable_offload: False

  model:
    model_name: "openpi"
    precision: null
    num_action_chunks: 5 # interface for the env
    action_dim: 7
    is_lora: False
    lora_rank: 2
    gradient_checkpointing: False
    use_wrist_image: True
    use_proprio: True
    num_steps: 4
    # openpi specific parameters
    openpi:
      noise_level: 0.5
      action_chunk: ${actor.model.num_action_chunks}
      num_steps: ${actor.model.num_steps}
      train_expert_only: True
      action_env_dim: ${actor.model.action_dim}
      noise_method: "flow_sde" 
      add_value_head: False
      detach_critic_input: False

  optim:
    lr: 5.0e-6
    value_lr: 1.0e-4
    adam_beta1: 0.9
    adam_beta2: 0.95
    adam_eps: 1.0e-05
    clip_grad: 1.0

  fsdp_config:                  # below are fsdp configs, for more details, see https://pytorch.org/docs/stable/fsdp.html and https://huggingface.co/docs/accelerate/package_reference/fsdp

    strategy: "fsdp"            # FSDP strategy: ["fsdp", "fsdp2"]

    # Sharding strategy: "full_shard" (shard parameters, gradients, optimizer states),
    # "shard_grad_op" (shard gradients and optimizer states only), 
    # "hybrid_shard" (combines data parallelism and model parallelism - sharding within FSDP groups, replicate across DDP groups),
    # "no_shard" (no sharding)
    sharding_strategy: "no_shard"

    cpu_offload: True           # whether to offload parameters and gradients to CPU when not in use，if True, actor's enable_offload should be True too.
    offload_pin_memory: False    # whether FSDP2's CPU offload policy should pin memory when cpu_offload is True
    reshard_after_forward: True  # if True, FSDP2 will reshard parameters after forward pass to save memory

    enable_gradient_accumulation: True  # if True, gradient accumulation will be enabled for FSDP/FSDP2 training, which will enhance training performance but increase memory usage.
    forward_prefetch: False             # if True, FSDP will explicitly prefetches the next upcoming all-gather while executing in the forward pass. only use with static graphs. Overlaps communication with computation to improve performance.
    limit_all_gathers: False            # if True, FSDP will synchronizes CPU threads to limit the number of concurrent all-gathers. Only affects strategies that schedule all-gathers
    backward_prefetch: null             # options are null, 'pre', 'post'. if 'pre', FSDP will prefetch the next upcoming all-gather while computing gradients. if 'post', FSDP will prefetch the next all-gather until current gradient is computed.
    use_orig_params: False              # if True, FSDP will use module's original parameters, it means it will expose nn.Module.named_parameters rather than FlatParameter
    use_liger_kernel: False             # if True, liger_kernel will be used for FSDP, note that currently supported models in RLinf are [qwen2.5, qwen2.5-vl], for more details, see liger_kernel's doc

    fsdp_size: -1               # Number of GPUs per FSDP group for hybrid sharding. -1 means use all available GPUs in a single FSDP group

    mixed_precision:            # mixed precision settings for fsdp/fsdp2
      param_dtype: ${actor.model.precision}
      reduce_dtype: ${actor.model.precision}
      buffer_dtype: ${actor.model.precision}

    amp:
      enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
      precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
      use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.

reward:
  use_reward_model: False

critic:
  use_critic_model: False

> have u changed any config in yaml？

**Comment by WinstonWmj** [comment]
@airrain Sorry about the delay. I think the `NaN` problem is probably caused by the small number of `group_size`, try increase it, we recommand `group_size=8`.
